File: src/nonlinear/nonlinear_factor_graph.rs
use cxx::UniquePtr;

use crate::{
    base::vector::Vector3, geometry::pose3::Pose3, imu::imu_bias::ConstantBias, inference::key::IntoKey, linear::noise_model::{DiagonalNoiseModel, IsotropicNoiseModel}, navigation::combined_imu_factor::CombinedImuFactor, slam::projection_factor::SmartProjectionPoseFactorCal3S2
};

pub struct NonlinearFactorGraph {
    pub(super) inner: UniquePtr<::sys::NonlinearFactorGraph>,
}

impl Default for NonlinearFactorGraph {
    fn default() -> Self {
        Self {
            inner: ::sys::default_nonlinear_factor_graph(),
        }
    }
}

impl NonlinearFactorGraph {
    pub fn resize(&mut self, size: usize) {
        ::sys::nonlinear_factor_graph_resize(self.inner.pin_mut(), size)
    }

    pub fn add_between_factor_pose3(
        &mut self,
        from: impl IntoKey,
        to: impl IntoKey,
        measured: &Pose3,
        model: &DiagonalNoiseModel,
    ) {
        ::sys::nonlinear_factor_graph_add_between_factor_pose3(
            self.inner.pin_mut(),
            from.into_key(),
            to.into_key(),
            &measured.inner,
            &model.to_base_model().inner,
        )
    }

    pub fn add_prior_factor_pose3(
        &mut self,
        symbol: impl IntoKey,
        prior: &Pose3,
        model: &DiagonalNoiseModel,
    ) {
        ::sys::nonlinear_factor_graph_add_prior_factor_pose3(
            self.inner.pin_mut(),
            symbol.into_key(),
            &prior.inner,
            &model.to_base_model().inner,
        )
    }

    pub fn add_prior_factor_constant_bias(
        &mut self,
        symbol: impl IntoKey,
        prior: &ConstantBias,
        model: &IsotropicNoiseModel,
    ) {
        ::sys::nonlinear_factor_graph_add_prior_factor_constant_bias(
            self.inner.pin_mut(),
            symbol.into_key(),
            &prior.inner,
            &model.to_base_model().inner,
        )
    }

    pub fn add_prior_factor_vector3(
        &mut self,
        symbol: impl IntoKey,
        prior: &Vector3,
        model: &IsotropicNoiseModel,
    ) {
        ::sys::nonlinear_factor_graph_add_prior_factor_vector3(
            self.inner.pin_mut(),
            symbol.into_key(),
            &prior.inner,
            &model.to_base_model().inner,
        )
    }

    pub fn add_combined_imu_factor(
        &mut self,
        factor: &CombinedImuFactor,
    ) {
        ::sys::nonlinear_factor_graph_add_combined_imu_factor(
            self.inner.pin_mut(),
            &factor.inner,
        )
    }

    pub fn add_smartfactor(&mut self, factor: &SmartProjectionPoseFactorCal3S2) {
        ::sys::nonlinear_factor_graph_add_smart_projection_pose_factor(self.inner.pin_mut(), &factor.inner)
    }

}

File: sys/src/nonlinear/nonlinear_factor_graph.cc
#include "nonlinear_factor_graph.h"
#include "../base/rust.hpp"

namespace gtsam {

std::unique_ptr<NonlinearFactorGraph> default_nonlinear_factor_graph() {
  return std::make_unique<NonlinearFactorGraph>();
}

void nonlinear_factor_graph_resize(NonlinearFactorGraph &graph, size_t size)
{
    return graph.resize(size);
}

void nonlinear_factor_graph_add_between_factor_pose3(
    NonlinearFactorGraph &graph, Key key1, Key key2, const Pose3 &measured,
    const std::shared_ptr<noiseModel::Base> &model) {
  return graph.add(
      BetweenFactorPose3(key1, key2, measured, to_boost_ptr(model)));
}

void nonlinear_factor_graph_add_prior_factor_pose3(
    NonlinearFactorGraph &graph, Key key, const Pose3 &prior,
    const std::shared_ptr<noiseModel::Base> &model) {
  return graph.addPrior(key, prior, to_boost_ptr(model));
}

void nonlinear_factor_graph_add_prior_factor_constant_bias(
    NonlinearFactorGraph &graph, Key key, const imuBias::ConstantBias &prior,
    const std::shared_ptr<noiseModel::Base> &model)
{
    return graph.addPrior(key, prior, to_boost_ptr(model));
}

void nonlinear_factor_graph_add_prior_factor_vector3(
    NonlinearFactorGraph &graph, Key key, const Vector3 &prior,
    const std::shared_ptr<noiseModel::Base> &model)
{
    return graph.addPrior(key, prior, to_boost_ptr(model));
}

void nonlinear_factor_graph_add_combined_imu_factor(
    NonlinearFactorGraph &graph, const std::shared_ptr<CombinedImuFactor> &factor) {
    return graph.add(to_boost_ptr(factor));
}

void nonlinear_factor_graph_add_smart_projection_pose_factor(
    NonlinearFactorGraph &graph,
    const gtsam::SmartProjectionPoseFactor<gtsam::Cal3_S2> &factor)
{
    return graph.add(factor);
}

} // namespace gtsam
